pathd/path_ted.c

```diff
@@ -0,0 +1,693 @@
+		break;
+
+	default:
+		zlog_debug(
+			"%s: [rcv ted] TED received unknown message type [%d]",
+			__func__, msg->type);
+		break;
+	}
+	return 0;
+}
+
+uint32_t path_ted_query_type_f(struct ipaddr *local, struct ipaddr *remote)
+{
+	uint32_t sid = MPLS_LABEL_NONE;
+	struct ls_edge *edge;
+
+	if (!path_ted_is_initialized())
+		return MPLS_LABEL_NONE;
+
+	switch (local->ipa_type) {
+	case IPADDR_V4:
+		/* We have local and remote ip */
+		/* so check all attributes in ted */
+		frr_each (edges, &ted_state_g.ted->edges, edge) {
+			if (edge->attributes->standard.local.s_addr
+				    == local->ip._v4_addr.s_addr
+			    && edge->attributes->standard.remote.s_addr
+				       == remote->ip._v4_addr.s_addr
+			    && CHECK_FLAG(edge->attributes->flags,
+					  LS_ATTR_ADJ_SID)) {
+				sid = edge->attributes->adj_sid[0]
+					      .sid; /* from primary */
+				break;
+			}
+		}
+		break;
+	case IPADDR_V6:
+		frr_each (edges, &ted_state_g.ted->edges, edge) {
+			if (memcmp(&edge->attributes->standard.local6,
+				   &local->ip._v6_addr,
+				   sizeof(local->ip._v6_addr))
+			    && (memcmp(&edge->attributes->standard.remote6,
+				       &remote->ip._v6_addr,
+				       sizeof(remote->ip._v6_addr))
+				&& CHECK_FLAG(edge->attributes->flags,
+					      LS_ATTR_ADJ_SID))) {
+				sid = edge->attributes->adj_sid[0]
+					      .sid; /* from primary */
+				break;
+			}
+		}
+		break;
+	case IPADDR_NONE:
+		break;
+	}
+
+	return sid;
+}
+
+uint32_t path_ted_query_type_c(struct prefix *prefix, uint8_t algo)
+{
+	uint32_t sid = MPLS_LABEL_NONE;
+	struct ls_subnet *subnet;
+
+	if (!path_ted_is_initialized())
+		return MPLS_LABEL_NONE;
+
+	switch (prefix->family) {
+	case AF_INET:
+	case AF_INET6:
+		frr_each (subnets, &ted_state_g.ted->subnets, subnet) {
+			if (prefix_same(&subnet->ls_pref->pref, prefix)
+			    && (CHECK_FLAG(subnet->ls_pref->flags, LS_PREF_SR))
+			    && (subnet->ls_pref->sr.algo == algo))
+				sid = subnet->ls_pref->sr.sid;
+		}
+		break;
+	default:
+		break;
+	}
+
+	return sid;
+}
+
+uint32_t path_ted_query_type_e(struct prefix *prefix, uint32_t iface_id)
+{
+	uint32_t sid = MPLS_LABEL_NONE;
+	struct ls_subnet *subnet;
+	struct listnode *lst_node;
+	struct ls_edge *edge;
+
+	if (!path_ted_is_initialized())
+		return MPLS_LABEL_NONE;
+
+	switch (prefix->family) {
+	case AF_INET:
+	case AF_INET6:
+		frr_each (subnets, &ted_state_g.ted->subnets, subnet) {
+			if (prefix_same(&subnet->ls_pref->pref, prefix)) {
+				/* from the vertex linked in subnet */
+				/* loop over outgoing edges */
+				for (ALL_LIST_ELEMENTS_RO(
+					     subnet->vertex->outgoing_edges,
+					     lst_node, edge)) {
+					/* and look for ifaceid */
+					/* so get sid of attribute */
+					if (CHECK_FLAG(edge->attributes->flags,
+						       LS_ATTR_LOCAL_ID)
+					    && edge->attributes->standard
+							       .local_id
+						       == iface_id) {
+						sid = subnet->ls_pref->sr.sid;
+						break;
+					}
+				}
```

Comment:
Here; for readability, and speed of performance, break once you find the subnet and then loop on vertex edges after the frr_each